internal/journal/omap.go

```diff
@@ -0,0 +1,138 @@
+package journal
+
+import (
+	"context"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	"github.com/ceph/go-ceph/rados"
+	"k8s.io/klog"
+)
+
+// listExcess is the number of false-positive key-value pairs we will
+// accept from ceph when getting omap values.
+var listExcess = 32
+
+func getOMapValues(
+	ctx context.Context,
+	conn *Connection,
+	poolName, namespace, oid, prefix string, keys []string) (map[string]string, error) {
+	// fetch and configure the rados ioctx
+	ioctx, err := conn.conn.GetIoctx(poolName)
+	if err != nil {
+		return nil, omapPoolError(poolName, err)
+	}
+	defer ioctx.Destroy()
+
+	if namespace != "" {
+		ioctx.SetNamespace(namespace)
+	}
+
+	results := map[string]string{}
+	// want is our "lookup map" that ensures O(1) checks for keys
+	// while iterating, without needing to complicate the caller.
+	want := make(map[string]bool, len(keys))
+	for i := range keys {
+		want[keys[i]] = true
+	}
+
+	err = ioctx.ListOmapValues(
+		oid, "", prefix, int64(len(want)+listExcess),
+		func(key string, value []byte) {
+			if want[key] {
+				results[key] = string(value)
+			}
+		},
+	)
+	switch err {
+	case nil:
+	case rados.ErrNotFound:
+		klog.Errorf(
```

Comment:
can we log which key not found? it would be useful for debugging

Reply:
We can log the keys requested, but logging keys-not-found doesn't make sense for this case because it is only returned if the oid is not found. The only reason it returns ErrKeyNotFound is for backwards compat. with the previous implementation.
So I'm not sure it would really be helpful.